lib/oauthdb.js

```diff
@@ -0,0 +1,177 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+/* Operations on OAuth database state.
+ *
+ * Currently this is not actually talking to a database,
+ * it's making authenticated calls to the fxa-oauth-server API
+ * to interrogate and manipulate its state, essentially treating
+ * fxa-oauth-server as a kind of backend micro-service.
+ *
+ * We want to work towards merging the fxa-oauth-server code
+ * directly into the main fxa-auth-server process, at which point
+ * this abstraction will convert into more direct db access.
+ *
+ */
+
+const Joi = require('joi')
+const JWT = require('jsonwebtoken')
+
+const P = require('./promise')
+const createBackendServiceAPI = require('./backendService')
+const error = require('./error')
+const validators = require('./routes/validators')
+
+// The oauth-server's error numbers overlap and conflict
+// with the auth-server's, so we have to map them to new ones.
+function mapOAuthError(err) {
+  if (err instanceof error) {
+    return err
+  }
+  switch (err.errno) {
+    case 101:
+      return error.unknownClientId(err.clientId)
+    case 119:
+      return error.staleAuthAt(err.authAt)
+    default:
+      return error.unexpectedError()
```

Comment:
related to Shane's comment, is there anything we can do here to get more info? Some extra log to determine what is happening in case of `unexpectedError`s?

Reply:
Good idea, I'll add a `log.warn` here with details of the underlying error.